app/controllers/course_summaries_controller.rb

```diff
@@ -15,4 +15,113 @@ def populate
   def get_marking_scheme_details
     redirect_to url_for(controller: 'marking_schemes', action: 'populate')
   end
+
+  def download_csv_grades_report
+    assignments = Assignment.order(:id)
+    csv_string = CSV.generate do |csv|
+      #Populates the entire CSV with headers, students and their marks
+      populated_csv = populate_students_and_marks(csv)
+    end
+    course_name = "#{COURSE_NAME}"
+    course_name_underscore = course_name.squish.downcase.tr(" ", "_")
+    send_data csv_string, disposition: 'attachment',
+              filename: "#{course_name_underscore}_grades_report.csv"
+  end
+end
+
+#---- HELPER FUNCTIONS FOR DOWNLOAD_CSV_GRADES_REPORT BELOW-----#
+
+#Helper function for download_csv_grades_report. Populates the headers in the CSV
+def populate_header_titles(assignments)
+  header = ['Username']
+  assignments.each do |assignment|
+    header.push(assignment.short_identifier)
+  end
+  header.push('Total marks')
+end
+
+#Helper function for populate_students_and_marks. Checks grouping, submission and it's mark for nil.
+def fill_row_with_marks(grouping, row)
+  if check_for_nil_fields(grouping)
+    row.push('')
+  else
+    row.push(grouping.current_submission_used.get_latest_result.total_mark / grouping.assignment.total_mark * 100)
+  end
+  row
+end
+
+#Helper function for download_csv_grades_report. Loads up the CSV with headers, students and marks.
+def populate_students_and_marks(csv)
+  students = Student.all
+  assignments = Assignment.order(:id)
+  #Populates the csv header titles here
+  csv << populate_header_titles(assignments)
+  students.each do |student|
+    total_mark = 0
+    row = []
+    row.push(student.user_name)
+      assignments.each do |assignment|
+        grouping = student.accepted_grouping_for(assignment.id)
+        #Check for nil fields. If not nil, then fill up the rows!
+        row = fill_row_with_marks(grouping, row)
+        total_mark += get_total_marks_for_student(grouping)
+      end
+    row.push(total_mark)
+    csv << row
+  end
+  csv
+end
+
+#Return the total marks for that student with its appropriate weights
+def get_total_marks_for_student(grouping)
+  current_total_marks = 0
+  if not check_for_nil_fields(grouping)
+  #Create Hashmap with Criteria ID as keys and it's weight as values
```

Comment:
Incorrect indentation detected (column 2 instead of 4).<br>Missing space after #.